File: agent/tools/workspace/workspace_manager.py
from __future__ import annotations

import os
from pathlib import Path

from tools.workspace.BinaryFileError import BinaryFileError
from tools.workspace.DepthLimitError import DepthLimitError
from tools.workspace.FileCountLimitError import FileCountLimitError
from tools.workspace.FileSizeLimitError import FileSizeLimitError
from tools.workspace.PathTraversalError import PathTraversalError
from tools.workspace.WorkspaceError import WorkspaceError
from tools.workspace.workspace_limits import (
    CACHE_DIR,
    MAX_FILE_SIZE,
    OPERATION_TIMEOUT,
)
from tools.workspace.workspace_listing import build_workspace_listing
from tools.workspace.workspace_path_guard import WorkspacePathGuard
from tools.workspace.workspace_stats import build_workspace_stats


class WorkspaceManager:
    """Workspace 核心管理类"""

    def __init__(self, base_path: str | None = None) -> None:
        if base_path is None:
            base_path = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), "workspace")
        self._base_path = Path(base_path).resolve()
        self._path_guard = WorkspacePathGuard(self._base_path)

    def _get_session_path(self, user_id: int | None, session_id: int | None) -> Path:
        """获取用户 session 的 workspace 路径"""
        return self._path_guard.get_session_path(user_id, session_id)

    def _validate_path(self, user_id: int | None, session_id: int | None, relative_path: str) -> Path:
        """验证路径安全性，返回绝对路径"""
        return self._path_guard.validate_path(user_id, session_id, relative_path)

    def _ensure_session_path(self, user_id: int | None, session_id: int | None) -> Path:
        """确保 session 路径存在"""
        return self._path_guard.ensure_session_path(user_id, session_id)

    def _check_depth(self, user_id: int | None, session_id: int | None, path: Path) -> None:
        """检查目录深度"""
        self._path_guard.check_depth(user_id, session_id, path)

    def _check_file_limit(self, session_path: Path) -> None:
        """检查文件数量限制"""
        self._path_guard.check_file_limit(session_path)

    def read(self, user_id: int | None, session_id: int | None, relative_path: str, offset: int = 0, limit: int = 8192) -> str:
        """读取文件内容"""
        path = self._validate_path(user_id, session_id, relative_path)
        if not path.exists() or not path.is_file():
            raise FileNotFoundError(f"文件不存在: {relative_path}")

        # 检查文件大小
        file_size = path.stat().st_size
        if file_size > MAX_FILE_SIZE:
            raise FileSizeLimitError(f"文件过大（最大 {MAX_FILE_SIZE} 字节）: {file_size}")

        # 限制读取范围
        limit = min(limit, MAX_FILE_SIZE)

        data = path.read_bytes()
        return data[offset : offset + limit].decode("utf-8", errors="replace")

    def write(self, user_id: int | None, session_id: int | None, relative_path: str, content: str, is_final: bool = False) -> dict:
        """写入文件内容"""
        session_path = self._ensure_session_path(user_id, session_id)
        path = self._validate_path(user_id, session_id, relative_path)

        target_path = path if not is_final else self._path_guard.final_path(session_path, relative_path)
        self._check_depth(user_id, session_id, target_path)

        # 检查文件内容大小
        content_bytes = content.encode("utf-8")
        if len(content_bytes) > MAX_FILE_SIZE:
            raise FileSizeLimitError(f"内容过大（最大 {MAX_FILE_SIZE} 字节）")

        self._check_file_limit(session_path)

        # 确保父目录存在
        path.parent.mkdir(parents=True, exist_ok=True)

        if is_final:
            # 写入 final 目录
            final_path = self._path_guard.final_path(session_path, relative_path)
            final_path.parent.mkdir(parents=True, exist_ok=True)
            with open(final_path, "w", encoding="utf-8") as f:
                f.write(content)
            return {"path": str(final_path.relative_to(session_path)), "size": len(content_bytes)}
        else:
            with open(path, "w", encoding="utf-8") as f:
                f.write(content)
            return {"path": relative_path, "size": len(content_bytes)}

    def edit(self, user_id: int | None, session_id: int | None, relative_path: str, old_string: str, new_string: str, is_final: bool = False) -> dict:
        """编辑文件内容"""
        path = self._validate_path(user_id, session_id, relative_path)
        if not path.exists() or not path.is_file():
            raise FileNotFoundError(f"文件不存在: {relative_path}")

        with open(path, "r", encoding="utf-8") as f:
            content = f.read()

        if old_string not in content:
            raise ValueError(f"未找到要替换的内容: {old_string[:50]}...")

        new_content = content.replace(old_string, new_string, 1)

        if is_final:
            session_path = self._get_session_path(user_id, session_id)
            final_path = self._path_guard.final_path(session_path, relative_path)
            final_path.parent.mkdir(parents=True, exist_ok=True)
            with open(final_path, "w", encoding="utf-8") as f:
                f.write(new_content)
            return {"path": str(final_path.relative_to(session_path)), "replaced": True}
        else:
            with open(path, "w", encoding="utf-8") as f:
                f.write(new_content)
            return {"path": relative_path, "replaced": True}

    def list(self, user_id: int | None, session_id: int | None, relative_path: str = ".", recursive: bool = False) -> list[dict]:
        """列出目录内容"""
        session_path = self._ensure_session_path(user_id, session_id)
        path = self._validate_path(user_id, session_id, relative_path)

        return build_workspace_listing(path, recursive=recursive)

    def create_dir(self, user_id: int | None, session_id: int | None, relative_path: str, is_final: bool = False) -> dict:
        """创建目录"""
        session_path = self._ensure_session_path(user_id, session_id)
        path = self._validate_path(user_id, session_id, relative_path)

        target_path = path if not is_final else self._path_guard.final_path(session_path, relative_path)
        self._check_depth(user_id, session_id, target_path)

        if is_final:
            final_path = self._path_guard.final_path(session_path, relative_path)
            final_path.mkdir(parents=True, exist_ok=True)
            return {"path": str(final_path.relative_to(session_path)), "created": True}
        else:
            path.mkdir(parents=True, exist_ok=True)
            return {"path": relative_path, "created": True}

    def cleanup_cache(self, user_id: int | None, session_id: int | None) -> dict:
        """清理缓存目录"""
        session_path = self._get_session_path(user_id, session_id)
        cache_path = session_path / CACHE_DIR

        cleaned_files = 0
        cleaned_size = 0

        if cache_path.exists():
            for item in cache_path.rglob("*"):
                if item.is_file():
                    cleaned_size += item.stat().st_size
                    item.unlink()
                    cleaned_files += 1
            # 删除空目录
            for item in sorted(cache_path.rglob("*"), key=lambda x: len(x.parts), reverse=True):
                if item.is_dir() and not any(item.iterdir()):
                    item.rmdir()

        return {
            "cleaned_files": cleaned_files,
            "cleaned_size": cleaned_size,
        }

    def get_stats(self, user_id: int | None, session_id: int | None) -> dict:
        """获取 workspace 统计信息"""
        session_path = self._get_session_path(user_id, session_id)
        return build_workspace_stats(session_path, user_id, session_id)

File: agent-ts/src/WorkspaceManager.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import { CACHE_DIR, FINAL_DIR, MAX_FILE_SIZE } from "./WorkspaceLimits.js";
import type { WorkspaceListing } from "./WorkspaceListing.js";
import { WorkspaceFileSystem } from "./WorkspaceFileSystem.js";
import { WorkspaceListingBuilder } from "./WorkspaceListingBuilder.js";
import { WorkspacePathGuard } from "./WorkspacePathGuard.js";
import { WorkspaceError } from "./WorkspaceError.js";
import type { WorkspaceStats } from "./WorkspaceStats.js";
import { WorkspaceStatsCollector } from "./WorkspaceStatsCollector.js";

export class WorkspaceManager {
  private readonly fileSystem = new WorkspaceFileSystem();
  private readonly listingBuilder = new WorkspaceListingBuilder(this.fileSystem);
  private readonly statsCollector = new WorkspaceStatsCollector(this.fileSystem);
  private readonly pathGuard: WorkspacePathGuard;
  private readonly basePath: string;

  constructor(basePath: string) {
    this.basePath = path.resolve(basePath);
    this.pathGuard = new WorkspacePathGuard(this.basePath);
  }

  async read(userId: number | null, sessionId: number | null, relativePath: string, offset = 0, limit = 8192): Promise<string> {
    const targetPath = this.pathGuard.validatePath(userId, sessionId, relativePath);
    const stat = await fs.stat(targetPath);
    if (!stat.isFile()) {
      throw new WorkspaceError(`文件不存在: ${relativePath}`);
    }
    if (stat.size > MAX_FILE_SIZE) {
      throw new WorkspaceError(`文件过大（最大 ${MAX_FILE_SIZE} 字节）: ${stat.size}`);
    }

    const content = await fs.readFile(targetPath, "utf8");
    return content.slice(offset, offset + Math.min(limit, MAX_FILE_SIZE));
  }

  async write(
    userId: number | null,
    sessionId: number | null,
    relativePath: string,
    content: string,
    isFinal = false
  ): Promise<{ path: string; size: number }> {
    const sessionPath = await this.ensureSessionPath(userId, sessionId);
    const normalPath = this.pathGuard.validatePath(userId, sessionId, relativePath);
    const targetPath = isFinal ? this.pathGuard.finalPath(sessionPath, relativePath) : normalPath;
    this.pathGuard.checkDepth(sessionPath, targetPath);

    const contentBytes = Buffer.byteLength(content, "utf8");
    if (contentBytes > MAX_FILE_SIZE) {
      throw new WorkspaceError(`内容过大（最大 ${MAX_FILE_SIZE} 字节）`);
    }
    this.pathGuard.checkFileLimit(await this.countWorkingFiles(sessionPath));

    await fs.mkdir(path.dirname(targetPath), { recursive: true });
    await fs.writeFile(targetPath, content, "utf8");
    return { path: path.relative(sessionPath, targetPath), size: contentBytes };
  }

  async edit(
    userId: number | null,
    sessionId: number | null,
    relativePath: string,
    oldString: string,
    newString: string,
    isFinal = false
  ): Promise<{ path: string; replaced: boolean }> {
    const sessionPath = this.pathGuard.getSessionPath(userId, sessionId);
    const normalPath = this.pathGuard.validatePath(userId, sessionId, relativePath);
    const content = await fs.readFile(normalPath, "utf8");
    if (!content.includes(oldString)) {
      throw new WorkspaceError(`未找到要替换的内容: ${oldString.slice(0, 50)}...`);
    }

    const targetPath = isFinal ? this.pathGuard.finalPath(sessionPath, relativePath) : normalPath;
    await fs.mkdir(path.dirname(targetPath), { recursive: true });
    await fs.writeFile(targetPath, content.replace(oldString, newString), "utf8");
    return { path: path.relative(sessionPath, targetPath), replaced: true };
  }

  async list(userId: number | null, sessionId: number | null, relativePath = ".", recursive = false): Promise<WorkspaceListing[]> {
    await this.ensureSessionPath(userId, sessionId);
    const targetPath = this.pathGuard.validatePath(userId, sessionId, relativePath);
    return this.listingBuilder.build(targetPath, recursive);
  }

  async createDir(
    userId: number | null,
    sessionId: number | null,
    relativePath: string,
    isFinal = false
  ): Promise<{ path: string; created: boolean }> {
    const sessionPath = await this.ensureSessionPath(userId, sessionId);
    const normalPath = this.pathGuard.validatePath(userId, sessionId, relativePath);
    const targetPath = isFinal ? this.pathGuard.finalPath(sessionPath, relativePath) : normalPath;
    this.pathGuard.checkDepth(sessionPath, targetPath);

    await fs.mkdir(targetPath, { recursive: true });
    return { path: path.relative(sessionPath, targetPath), created: true };
  }

  async cleanupCache(userId: number | null, sessionId: number | null): Promise<{ cleaned_files: number; cleaned_size: number }> {
    const sessionPath = this.pathGuard.getSessionPath(userId, sessionId);
    const cachePath = path.join(sessionPath, CACHE_DIR);
    const stats = { cleaned_files: 0, cleaned_size: 0 };
    if (!(await this.fileSystem.exists(cachePath))) {
      return stats;
    }

    for (const filePath of await this.fileSystem.walk(cachePath)) {
      const stat = await fs.stat(filePath);
      if (!stat.isFile()) {
        continue;
      }
      stats.cleaned_files += 1;
      stats.cleaned_size += stat.size;
      await fs.unlink(filePath);
    }
    await fs.rm(cachePath, { recursive: true, force: true });
    return stats;
  }

  async getStats(userId: number | null, sessionId: number | null): Promise<WorkspaceStats> {
    const sessionPath = this.pathGuard.getSessionPath(userId, sessionId);
    return this.statsCollector.collect(sessionPath, userId, sessionId);
  }

  private async ensureSessionPath(userId: number | null, sessionId: number | null): Promise<string> {
    const sessionPath = this.pathGuard.getSessionPath(userId, sessionId);
    await fs.mkdir(sessionPath, { recursive: true });
    return sessionPath;
  }

  private async countWorkingFiles(sessionPath: string): Promise<number> {
    if (!(await this.fileSystem.exists(sessionPath))) {
      return 0;
    }
    let count = 0;
    for (const filePath of await this.fileSystem.walk(sessionPath)) {
      const parts = filePath.split(path.sep);
      if (!parts.includes(CACHE_DIR) && !parts.includes(FINAL_DIR) && (await fs.stat(filePath)).isFile()) {
        count += 1;
      }
    }
    return count;
  }

}
